group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java

```diff
@@ -266,9 +295,21 @@ public CompletableFuture<JoinGroupResponseData> joinGroup(
             return FutureUtils.failedFuture(Errors.COORDINATOR_NOT_AVAILABLE.exception());
         }
 
-        return FutureUtils.failedFuture(Errors.UNSUPPORTED_VERSION.exception(
-            "This API is not implemented yet."
-        ));
+        CompletableFuture<JoinGroupResponseData> responseFuture = new CompletableFuture<>();
+
+        if (!isValidGroupId(request.groupId(), ApiKeys.forId(request.apiKey()))) {
+            responseFuture.complete(new JoinGroupResponseData()
+                .setMemberId(request.memberId())
+                .setErrorCode(Errors.INVALID_GROUP_ID.code()));
+
+            return responseFuture;
+        }
+
+        runtime.scheduleGenericGroupOperation("generic-group-join",
+            topicPartitionFor(request.groupId()),
+            coordinator -> coordinator.genericGroupJoin(context, request, responseFuture));
+
+        return responseFuture;
```

Comment:
We probably need to convert some of the exceptions like I did for the consumer group heartbeat request.

Reply:
i'm a bit confused, is the coordinator write operation future the result we wait for committing?

Reply:
i don't think this is the right place; we need to add the logic inside group metadata manager. i have done so in the sync PR.

the reason is that for generic group apis, the append future is what we want the logic for when handling log append/commit errors.

whereas for the new protocol, the consumer group heartbeat waits to return the results from the append/commit.